File: techstore-frontend/techstore-admin/src/app/features/warehouse/warehouse.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse } from '../../shared/models/api-response.model';
import { WarehouseResponse } from './models/warehouse.model';
import { WarehouseCreateRequest, WarehouseUpdateRequest } from './models/warehouse-request.model';

@Injectable({ providedIn: 'root' })
export class WarehouseService {
  private baseUrl = environment.warehouseUrl + '/warehouses';

  constructor(private http: HttpClient) {}

  create(req: WarehouseCreateRequest):
    Observable<ApiResponse<WarehouseResponse>> {
    return this.http.post<ApiResponse<WarehouseResponse>>(this.baseUrl, req);
  }

  update(id: number, req: WarehouseUpdateRequest):
    Observable<ApiResponse<WarehouseResponse>> {
    return this.http.put<ApiResponse<WarehouseResponse>>(`${this.baseUrl}/${id}`, req);
  }

  getById(id: number):
    Observable<ApiResponse<WarehouseResponse>> {
    return this.http.get<ApiResponse<WarehouseResponse>>(`${this.baseUrl}/${id}`);
  }

  getByName(name: string):
    Observable<ApiResponse<WarehouseResponse>> {
    return this.http.get<ApiResponse<WarehouseResponse>>(`${this.baseUrl}/name/${name}`);
  }

  getAll():
    Observable<ApiResponse<WarehouseResponse[]>> {
    return this.http.get<ApiResponse<WarehouseResponse[]>>(this.baseUrl);
  }

  getByStatus(status: string):
    Observable<ApiResponse<WarehouseResponse[]>> {
    return this.http.get<ApiResponse<WarehouseResponse[]>>(`${this.baseUrl}/status/${status}`);
  }

  getByAddress(addressId: string):
    Observable<ApiResponse<WarehouseResponse[]>> {
    return this.http.get<ApiResponse<WarehouseResponse[]>>(`${this.baseUrl}/address/${addressId}`);
  }

  delete(id: number):
    Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.baseUrl}/${id}`);
  }

  updateStatus(id: number, status: string):
    Observable<ApiResponse<WarehouseResponse>> {
    const params = new HttpParams().set('status', status);
    return this.http.put<ApiResponse<WarehouseResponse>>(`${this.baseUrl}/${id}/status`, null, { params });
  }
}

File: techstore-backend/warehouse-service/src/main/java/com/techstore/warehouse/service/WarehouseService.java
package com.techstore.warehouse.service;

import java.util.List;

import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.techstore.warehouse.constant.WarehouseStatus;
import com.techstore.warehouse.dto.request.WarehouseCreateRequest;
import com.techstore.warehouse.dto.request.WarehouseUpdateRequest;
import com.techstore.warehouse.dto.response.WarehouseResponse;
import com.techstore.warehouse.entity.Warehouse;
import com.techstore.warehouse.exception.AppException;
import com.techstore.warehouse.exception.ErrorCode;
import com.techstore.warehouse.mapper.WarehouseMapper;
import com.techstore.warehouse.repository.WarehouseRepository;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Service
@RequiredArgsConstructor
@Slf4j
public class WarehouseService {

    private final WarehouseRepository warehouseRepo;
    private final WarehouseMapper warehouseMapper;

    @PreAuthorize("hasAnyRole('ADMIN')")
    @Transactional
    public WarehouseResponse create(WarehouseCreateRequest req) {
        log.info("Creating warehouse with name: {}", req.getName());

        if (warehouseRepo.findByName(req.getName()).isPresent()) {
            throw new AppException(ErrorCode.WAREHOUSE_EXISTED);
        }

        Warehouse warehouse = warehouseMapper.toEntity(req);
        warehouse.setStatus(WarehouseStatus.ACTIVE.name());

        return warehouseMapper.toResponse(warehouseRepo.save(warehouse));
    }

    @PreAuthorize("hasAnyRole('ADMIN')")
    @Transactional
    public WarehouseResponse update(Long id, WarehouseUpdateRequest req) {
        log.info("Updating warehouse with id: {}", id);

        Warehouse warehouse =
                warehouseRepo.findById(id).orElseThrow(() -> new AppException(ErrorCode.WAREHOUSE_NOT_FOUND));

        if (req.getName() != null && !req.getName().equals(warehouse.getName())) {
            warehouseRepo.findByName(req.getName()).ifPresent(w -> {
                throw new AppException(ErrorCode.WAREHOUSE_EXISTED);
            });
        }

        warehouseMapper.updateEntityFromRequest(req, warehouse);
        return warehouseMapper.toResponse(warehouseRepo.save(warehouse));
    }

    @PreAuthorize("hasAnyRole('ADMIN','WAREHOUSE_STAFF')")
    public WarehouseResponse getById(Long id) {
        return warehouseMapper.toResponse(
                warehouseRepo.findById(id).orElseThrow(() -> new AppException(ErrorCode.WAREHOUSE_NOT_FOUND)));
    }

    @PreAuthorize("hasAnyRole('ADMIN','WAREHOUSE_STAFF')")
    public WarehouseResponse getByName(String name) {
        return warehouseMapper.toResponse(
                warehouseRepo.findByName(name).orElseThrow(() -> new AppException(ErrorCode.WAREHOUSE_NOT_FOUND)));
    }

    @PreAuthorize("hasAnyRole('ADMIN','WAREHOUSE_STAFF')")
    public List<WarehouseResponse> getAll() {
        return warehouseRepo.findAll().stream().map(warehouseMapper::toResponse).toList();
    }

    @PreAuthorize("hasAnyRole('ADMIN','WAREHOUSE_STAFF')")
    public List<WarehouseResponse> getByStatus(String status) {
        return warehouseRepo.findByStatus(status).stream()
                .map(warehouseMapper::toResponse)
                .toList();
    }

    @PreAuthorize("hasAnyRole('ADMIN','WAREHOUSE_STAFF')")
    public List<WarehouseResponse> getByAddress(String addressId) {
        return warehouseRepo.findByAddressId(addressId).stream()
                .map(warehouseMapper::toResponse)
                .toList();
    }

    @PreAuthorize("hasRole('ADMIN')")
    @Transactional
    public void delete(Long id) {
        log.info("Deleting warehouse with id: {}", id);

        Warehouse warehouse =
                warehouseRepo.findById(id).orElseThrow(() -> new AppException(ErrorCode.WAREHOUSE_NOT_FOUND));

        warehouseRepo.delete(warehouse);
    }

    @PreAuthorize("hasRole('ADMIN')")
    @Transactional
    public WarehouseResponse updateStatus(Long id, String status) {
        log.info("Updating warehouse status for id: {} to status: {}", id, status);

        Warehouse warehouse =
                warehouseRepo.findById(id).orElseThrow(() -> new AppException(ErrorCode.WAREHOUSE_NOT_FOUND));

        warehouse.setStatus(status);
        return warehouseMapper.toResponse(warehouseRepo.save(warehouse));
    }
}
